iroh-net/src/hp/portmapper.rs

```diff
@@ -14,41 +21,151 @@ pub struct ProbeResult {
     /// `Client::new` constructor (if any) will fire.
-    pub async fn get_cached_mapping_or_start_creating_one(&self) -> Option<SocketAddr> {
-        // TODO:
-        None
+    // TODO(@divagant-martian): fix docs. Also, this will probably only ever return ipv4 addresses.
+    pub async fn get_cached_mapping_or_start_creating_one(&mut self) -> Option<SocketAddr> {
+        // TODO(@divagant-martian): this has a lock ensure just one mapping attempt is underway
+        if let Some(mapping) = &mut self.current_mapping {
+            let now = Instant::now();
+            if now <= mapping.good_until() {
+                // TODO(@divagant-martian): this would go in a goroutine
+                if now >= mapping.renew_after() {
+                    if let Err(e) = mapping.renew().await {
+                        debug!("failed to renew port mapping {mapping}: {e}");
+                    }
+                }
+                // port mapping is still good regardless of renewal
+                Some(mapping.external().into())
+            } else {
+                // TODO(@divagant-martian): tailscale returns nil without clearing the mapping
+                None
+            }
+        } else {
+            match upnp::Mapping::new(SocketAddrV4::new(
+                std::net::Ipv4Addr::UNSPECIFIED,
+                self.local_port,
+            ))
+            .await
+            {
+                Ok(mapping) => {
+                    // TODO(@divagant-martian): too high? seems relevant to the user
+                    info!("upnp port mapping created {mapping}");
+                    // TODO(@divagant-martian): tailscale does not release the old mapping
+                    let external = mapping.external().into();
+                    self.current_mapping = Some(mapping);
+                    Some(external)
+                }
+                Err(e) => {
+                    debug!("Failed to create upnp port mapping: {e}");
+                    None
+                }
+            }
+        }
     }
 
-    pub fn have_mapping(&self) -> bool {
-        // TODO:
-        false
+    pub fn has_mapping(&self) -> bool {
+        self.current_mapping
+            .as_ref()
+            .filter(|mapping| Instant::now() <= mapping.good_until())
+            .is_some()
     }
 
-    pub fn note_network_down(&self) {
-        // TODO:
+    pub fn note_network_down(&mut self) {
+        self.current_mapping = None;
     }
 
-    pub fn close(&self) {
-        // TODO:
+    pub async fn close(&mut self) {
+        // TODO(@divagant-martian):
+        // tailscale has a `closed` bool, which is not very pretty.
+        // Easiest way to emulate this is consume self and have the owner have an Option. This does
+        // bring some complexity, so maybe we can just not close (permanently) but simply clean
+        // (invalidate mappings)
+        //
+        // The only place this is ever used is to prevent closing twice...
```

Comment:
Yeah, closing is not fully figured out, other parts also have the ability to close, but it seems kinda pointless as things can't be restarted. Ideally all the structures just consume `self`  or are able to be restarted.

Reply:
I don't have anything useful to say about this except to just complain about all the `closed` bools in the go code. I hate them everytime.